src/Symfony/Component/AssetMapper/ImportMap/ImportMapManager.php

```diff
@@ -0,0 +1,399 @@
+            }
+
+            $this->cleanupPackageFiles($currentEntries[$packageName]);
+            unset($currentEntries[$packageName]);
+        }
+
+        if ($update) {
+            foreach ($currentEntries as $importName => $entry) {
+                if (null === $entry->url) {
+                    continue;
+                }
+
+                // assume the import name === package name, unless we can parse
+                // the true package name from the URL
+                $packageName = $importName;
+                $registry = null;
+
+                // try to grab the package name & jspm "registry" from the URL
+                if (str_starts_with($entry->url, 'https://ga.jspm.io') && 1 === preg_match(self::PACKAGE_PATTERN, $entry->url, $matches)) {
+                    $packageName = $matches['package'];
+                    $registry = $matches['registry'] ?? null;
+                }
+
+                $packagesToRequire[] = new PackageRequireOptions(
+                    $packageName,
+                    null,
+                    $entry->isDownloaded,
+                    $entry->preload,
+                    $importName,
+                    $registry,
+                );
+
+                // remove it: then it will be re-added
+                $this->cleanupPackageFiles($entry);
+                unset($currentEntries[$importName]);
+            }
+        }
+
+        $newEntries = $this->requirePackages($packagesToRequire, $currentEntries);
+        $this->writeImportMapConfig($currentEntries);
+
+        return $newEntries;
+    }
+
+    /**
+     * Gets information about (and optionally downloads) the packages & updates the entries.
+     *
+     * Returns an array of the entries that were added.
+     *
+     * @param PackageRequireOptions[] $packagesToRequire
+     * @param array<string, ImportMapEntry> $importMapEntries
+     */
+    private function requirePackages(array $packagesToRequire, array &$importMapEntries): array
+    {
+        if (!$packagesToRequire) {
+            return [];
+        }
+
+        $installData = [];
+        $packageRequiresByName = [];
+        foreach ($packagesToRequire as $requireOptions) {
+            $constraint = $requireOptions->packageName;
+            if (null !== $requireOptions->versionConstraint) {
+                $constraint .= '@' . $requireOptions->versionConstraint;
+            }
+            if (null !== $requireOptions->registryName) {
+                $constraint = sprintf('%s:%s', $requireOptions->registryName, $constraint);
+            }
+            $installData[] = $constraint;
+            $packageRequiresByName[$requireOptions->packageName] = $requireOptions;
+        }
+
+        $json = [
+            'install' => $installData,
+            'flattenScope' => true,
+            // always grab production-ready assets
+            'env' => ['browser', 'module', 'production'],
```

Comment:
Shouldn't we do this only in production mode to ease debugging?

Reply:
I THINK what I have is correct, but tell me if you disagree. Try switching to `dev` mode and adding `react` in the builder - https://generator.jspm.io/ - you'll get `https://ga.jspm.io/npm:react@18.2.0/dev.index.js`. So even though we are adding `react` while coding in the `dev` environment, if we pass `dev` here, we'll get a CDN to the dev package. And then when the user deploys, they will still have the `dev` version.

So, if we wanted to be nicer here, I think we'd need to maintain both a `dev` and `prod` version of the URL inside of `importmap.php`. Not sure that's worth it.

Reply:
Makes sense! However having the ability to install the dev builds is a must have for debugging.

Reply:
I'll save this for a future PR - I think it'll involve making 2x requests to jspm - one for dev and one for prod - then maintain 2 URLs (when necessary) in `importmap.php`. I'm not aware of any packages that have a different dev vs prod that I would use, but I'm sure there are some (`react` is an example, but using react with importmaps seems edge-case).